### DISCUSSION TITLE
Linux

### DISCUSSION THREAD

**Discussion by p-robertobakker** [jimmejardine/qiqqa-open-source#325] [General]
Hi,
I have been an advocate for Qiqqa for a very long time. Also, I am an advocate FOSS, so I am happy that Qiqqa decided to go into FOSS. 
However! I am concerned that Qiqqa wil not, or after very long time, have an Linux version. Especially owing to the fact that I went completely from Windowns into Linux - only Qiqqa is a problem on Linux as Qiqqa is not working correctly in Wine. The Wine page mentions even 'Garbage' as review for Qiqqa in Wine. 
So, I hope very much that the Linux version will be developed. In other discussions I read it is not that easy, which I do understand. Patience I need ;)
Best regards, 
Roberto

**Comment by GerHobbelt** [comment]
Qiqqa uses several chunks of technology which must be replaced ("upgraded") for it to be viable on Linux.

Here's a few blurbs I jotted down that might be of interest:

- [Qiqqa Functionalities & Technology Areas](https://github.com/jimmejardine/qiqqa-open-source/blob/master/docs-src/Progress%20in%20Development/Qiqqa%20Functionalities%20%26%20Technology%20Areas.md)

- [Moving away for Windows-bound UI (WPF) to HTML :: feasibility tests with CEF / CEFSharp / CEFGlue / Chromely / WebView2](https://github.com/jimmejardine/qiqqa-open-source/blob/master/docs-src/Progress%20in%20Development/Moving%20away%20for%20Windows-bound%20UI%20(WPF)%20to%20HTML%20-%20feasibility%20tests%20with%20CEF%2BCEFSharp%2BCEFGlue%2BChromely.md)

It's all "stream of conscience" writing at the moment, but alas.

Anyway, Linux is definitely on my mind; it'll take a while to get there, though. Even if I did ignore the other areas, the entire UI must be replaced which is NOT-WPF for it to work on Linux/Wine.

**Reply by zhangxiaoxing** [reply]
Maybe we could rearrange the functions into sort of a "server" and a "UI/client" part, so porting to other platforms aside from Linux will also be benefited.

**Reply by GerHobbelt** [reply]
😄  Yup, that's my intent. The (long distance) target is to have the UI done in web technology (think 'electron' and that sort of thing: HTML pages + CSS/JS acting as your GUI, while the real work is done 'elsewhere': in C# (most of the current Qiqqa glue logic would still be there, I expect), while some parts are to be done in Java (SOLR/Lucene) and C/C++ (mupdf + tesseract = PDF rendering, text extraction and OCR; *possibly* also moving the database and sync functionality there (as there's some very hairy problems with NAS and sync that I'ld rather tackle in C/C++, "close to the metal", rather than C#) and *probably* also the LDA/... keyword extraction heuristics stuff that's currently done in C# 

> though the latter's reason to move is only good for large libraries and documents, where current C#/Qiqqa runs out of memory. Otherwise moving the LDA stuff is a mere 'hobby horse' 😉 for fun. I want to do some more meta-research dev there but that's falling between the cracks of kicking fundamental Qiqqa tech into the roaring twenties.

**Comment by GerHobbelt** [comment]
Cross-reference: issue #215

**Comment by p-robertobakker** [comment]
Hi jimmejardine,

Thank you for your reply and explanation. I am very happy that Linux is on your mind.
How can we help you?

Best wishes,
Roberto

8 mei 2021 01:04:48 Ger Hobbelt ***@***.***>:

> Qiqqa uses several chunks of technology which must be replaced ("upgraded") for it to be viable on Linux.
> 
> Here's a few blurbs I jotted down that might be of interest:
> 
* > Qiqqa Functionalities & Technology Areas[https://github.com/jimmejardine/qiqqa-open-source/blob/master/docs-src/Progress%20in%20Development/Qiqqa%20Functionalities%20%26%20Technology%20Areas.md]
> 
* > Moving away for Windows-bound UI (WPF) to HTML :: feasibility tests with CEF / CEFSharp / CEFGlue / Chromely / WebView2[https://github.com/jimmejardine/qiqqa-open-source/blob/master/docs-src/Progress%20in%20Development/Moving%20away%20for%20Windows-bound%20UI%20(WPF)%20to%20HTML%20-%20feasibility%20tests%20with%20CEF%2BCEFSharp%2BCEFGlue%2BChromely.md]

**Reply by GerHobbelt** [reply]
(Had to think about this for a bit; there's so much to do...)

# Designer

As Qiqqa *will* move to using HTML+CSS(+JS) as the tech behind a "cross platform" UI/UX, the UI needs to be recreated / rebuilt: that's part design, part web dev (CSS, JS/TS, ...).

If someone can do that or assist with that, super!

(By the way, ditto for the website. 😉  Less work for me there is more time to spend on Qiqqa components.)


# Sponsor

I'm considering how to set this up precisely, but the bottom line there is: when I get some sponsoring there, there's an even better reason to work on Qiqqa instead of doing other (paid) work.

I'm still a bit of two minds about this, though. Ethics and other complicated stuff. Call me crazy. 🤡😆 



# Documenter

I already have a **very kind** offer there. The man's been patiently waiting for me to get off my dev horse(s) for ages now. 😰 




# Developer

What would help me very much is checking, fixing and augmenting the build processes for the various (large) components as they start to appear. (Flagged as 🐧🐧🐧 below)

I'm Windows based nowadays for my Qiqqa-related development work, so having a few folks around who are Linux or Mac based and willing to do some dev work for the long haul would be a great boon: there's several experimental developments which need to be made workable on Linux + Mac before we can have a full-fledged Qiqqa for those platforms.

There's a couple of areas where work is pending or needs to be done:

- currently most of my time is spent in the MuPDF arena, where those tools are tweaked, extended and bulktested as I go. On Windows.

  That stuff is going to be the single major source of all data sourcing and page rendering of PDFs and related file formats. (think PDF download, PDF rendering for user viewing, PDF metadata extract, PDF text extraction, PDF OCR-ing, including customizable OCR pipelines)

  While I do my dev work using Visual Studio and thus MSVC projects and "solution files", there's also the Linux/Mac build chains to consider: Makefiles, CMake, autoconf/automake, depending on which part of that stratum you're looking at. That's not integrated nor tested yet: MuPDF comes with Makefiles, etc. but has been heavily worked on already so getting the Linux build chain back in order is on my "To Do, *Later*" list. 🐧🐧🐧
  Any help with that would be nice and would also be quite "independent" so there's minimal risk of getting merge collisions there.🐧🐧🐧

  The output of that would be commandline tools to process PDF files and at some point a local web server, which would be able to serve document-related requests.

  What I call 'MuPDF' is one of the **backends** of Qiqqa and a very important component.

- There are a couple of test applications I've been experimenting with to see if it's **viable** to move from current WPF/.NET UI (Windows-only; **not** my favorite design environment either 😅 ) to something more `electron`-like, but without the electron-node back-layer (as that's where we have C# already and I'm considering keeping a sizeable chunk of that for the "business logic layer", thus making the node layer in `electron` rather "in the way" as that would be more JavaScript shifting stuff around that would be done elsewhere (in the C# "business logic layer") -- no sense in rewriting everything in JS/TS without some added benefit.
  
  One of the GUI targets I have considered was/is [Chromely](https://chromely.net/): that **would be a very nice fit for "new age" Qiqqa**. It *would* be nice to know how easy it is to *build* and use on Linux in reality: what are actual developer experiences? 🐧
  *Unfortunately, I keep looking at their github page and see this: https://github.com/chromelyapps/Chromely and see dates telling me *at a glance* there's no large developer base active there. So I check their Insights>Network at https://github.com/chromelyapps/Chromely/network and I don't see a lot of "busy" either. Is it done/ready then? Not to my taste: I went to see if I could fiddle with it, to get a sense how easy or hard it would be to upgrade or patch **the most important component**: CEF (https://bitbucket.org/chromiumembedded/cef/src/master/) which **sensibly & happily** sees a metric ton of work done each month as it closely tracks the Chrome/Chromium browser.

  So I had a look around as Chromely, while very appealing, wasn't cooperating when I touched it. electron... NW.js (after having had a sniff at a few others)... None of it panned out: I want something with a large(ish) community and some or many developers, so that I can trust the *new one* to (a) probably be still around in 5 years, (b) be able to take the brunt of the effort tracking browser/web standards development off my hands for the duration, and (c) has hooks into the core so I can get at the download+cache pipeline (needed for Qiqqa Sniffer: PDF and BibTeX "automatic import"!) and the full DOM of the rendered web pages (again for Qiqqa Sniffer: metadata scraping for BibTeX et al)

  Long story shorter: either I do a "thin client" using CEF (or Microsoft WebView2 on Windows, as that's a very interesting equivalent for Windows at least 😄 ) and C/C++, which would thus be a purpose-built "dual head" web browser so we can have our security and Qiqqa Web Sniffer UX in the new world, or we do a WebView2/CEFGlue/Xilium.CEF C# based slightly thicker client, which would incorporate the "business logic" that's Qiqqa's current glue.

🐧🐧🐧 Question is: how easy and workable is that component on Linux and Mac? (build chain. debugging?! etc.) 🐧🐧🐧 so help there would be great.

----

## Summary

When we're talking help as in working on Qiqqa to make it real, there's two areas now or near future where work needs to done:

- 🐧🐧🐧 the new C/C++ backend: PDF/OCR/SQLite Sync/etc processing server component. (Which would largely replace current QiqqaOCR, pdfdraw, tesseract & SORAX tooling). 
  - 🐧🐧🐧 before there's this server, there's the commandline tools which serve to help test and doublecheck the efforts: those need to be built and tested on Linux and Mac. Help much appreciated.
- 🐧🐧🐧 the new GUI: test/create viability check application, which would, if successful, become the UI driver component, i.e. purpose-built browser which supports secured dual/multiple CEF instances for Qiqqa Web Sniffer UX: one browser to render UI+PDF, other SECURED instance in same screen which will act as live web browser, where we can hook into its data streams).  Help much appreciated.

**Reply by Waxwood** [reply]
Ten years ago I read about Qiqqa, but didn't try it because I was using Linux. For years I have been using Zotero. But today, I tried Qiqqa and am amazed by its functions- especially the ability to add note boxes to PDFs. 
My main concern is that Qiqqa is not cross platform and won't run in WINE. I am using a virtual machine Win10, but this is not a good long-term solution.

Do you have any time-frame for porting of Qiqqa to a new cross-platform architecture? 
I'll be very happy to donate to this project

**Reply by p-robertobakker** [reply]
Hi Everybody, 
I managed using Qiqqa with Wine, via:
[https://appdb.winehq.org/objectManager.php?sClass=version&iId=31023](url)
Look at **Manual Installation**
Then choose **.NET Framework 4.8:**
`wget 'https://download.visualstudio.microsoft.com/download/pr/7afca223-55d2-470a-8edc-6a1739ae3252/abd170b4b0ec15ad0222a809b761a036/ndp48-x86-x64-allos-enu.exe'`
`env WINEPREFIX=$HOME/winedotnet wine ~/ndp48-x86-x64-allos-enu.exe /q`

I hope it helps.

Best regards, 
Roberto

**Reply by derfolie** [reply]
Hi, I tried to install Qiqqa on Linux Mint via wine (v.10), but it doesn't work (I installed .net 4.8). Any specific advice, any additional steps you took? Thanks.

**Comment by p-robertobakker** [comment]
Hi Qiqqa community,
I just told to my PhD-students about Qiqqa, and they are interested. That's nice. Then, I checked again whether Qiqqa can be installed on Linux. It seems that this is still not possible, only with Wine etc. Is this correct? Or is the hope ;)?

**Comment by GerHobbelt** [comment]
Sorry,

Qiqqa currently doesn't exist for Linux (or Mac). Some folks have reported
successfully running Qiqqa under Wine, but from what I heard that takes
some savvy to get going, alas.

Met vriendelijke groeten / Best regards,

Ger Hobbelt

--------------------------------------------------
web:    http://www.hobbelt.com/
        http://www.hebbut.net/
mail:   ***@***.***
mobile: +31-6-11 120 978
--------------------------------------------------

**Reply by p-robertobakker** [reply]
How is Qiqqa and Linux at this moment? Would be nice if this was possible in the near future.
Best, Roberto
